pkg/quota/admission/podlimitrequest/admission.go

```diff
@@ -0,0 +1,160 @@
+}
+
+func ReadConfig(configFile io.Reader) (api.PodLimitRequestConfig, error) {
+	config := api.PodLimitRequestConfig{}
+	if configFile == nil || reflect.ValueOf(configFile).IsNil() /* pointer to nil */ {
+		glog.V(5).Infof("PodLimitRequest has no config to read.")
+		return config, nil
+	}
+	glog.V(5).Infof("PodLimitRequest about to read config:\n%v", configFile)
+	buffer := new(bytes.Buffer)
+	if _, err := buffer.ReadFrom(configFile); err != nil {
+		return config, err
+	}
+	err := yaml.Unmarshal(buffer.Bytes(), &config)
+	glog.V(5).Infof("PodLimitRequest config:\n%v\nerror: %v", config, err)
+	return config, err
+}
+
+func Validate(config api.PodLimitRequestConfig) error {
+	if config.Enabled {
+		if config.LimitCPUToMemoryRatio == 0.0 && config.CPURequestToLimitRatio == 0.0 && config.MemoryRequestToLimitRatio == 0.0 {
+			return fmt.Errorf("PodLimitRequest plugin enabled but no ratios specified")
+		}
+		if config.LimitCPUToMemoryRatio < 0.0 {
+			return fmt.Errorf("LimitCPUToMemoryRatio must be positive")
+		}
+		if config.CPURequestToLimitRatio < 0.0 || config.CPURequestToLimitRatio > 1.0 {
+			return fmt.Errorf("CPURequestToLimitRatio must be between 0.0 and 1.0")
+		}
+		if config.MemoryRequestToLimitRatio < 0.0 || config.MemoryRequestToLimitRatio > 1.0 {
+			return fmt.Errorf("MemoryRequestToLimitRatio must be between 0.0 and 1.0")
+		}
+	}
+	return nil
+}
+func (a *podLimitRequest) Validate() error {
+	if err := Validate(a.Config); err != nil {
+		return err
+	}
+	if a.ProjectCache == nil {
+		return fmt.Errorf("PodLimitRequest did not get a project cache")
+	}
+	return nil
+}
+
+// TODO this will need to update when we have pod requests/limits
+func (a *podLimitRequest) Admit(attr admission.Attributes) error {
+	glog.V(8).Infof("PodLimitRequest admission controller is invoked")
+	if !a.Config.Enabled || attr.GetResource() != kapi.Resource("pods") || attr.GetSubresource() != "" {
+		return nil // not applicable
+	}
+	pod, ok := attr.GetObject().(*kapi.Pod)
+	if !ok {
+		return admission.NewForbidden(attr, fmt.Errorf("unexpected object: %#v", attr.GetObject()))
+	}
+	glog.V(5).Infof("PodLimitRequest is looking at creating pod %s in project %s", pod.Name, attr.GetNamespace())
+
+	// allow annotations on project to override
+	if ns, err := a.ProjectCache.GetNamespace(attr.GetNamespace()); err != nil {
+		glog.Warningf("PodLimitRequest got an error retrieving namespace: %v", err)
+		return admission.NewForbidden(attr, err) // this should not happen though
+	} else {
+		projectEnabledPlugin, exists := ns.Annotations[podLimitRequestAnnotation]
+		if exists && projectEnabledPlugin != "true" {
+			glog.V(5).Infof("PodLimitRequest is disabled for project %s", attr.GetNamespace())
+			return nil // disabled for this project, do nothing
+		}
+	}
+
+	// Reuse LimitRanger logic to apply limit/req defaults from the project. Ignore validation
+	// errors, assume that LimitRanger will run after this plugin to validate.
+	a.LimitRanger.Admit(attr)
```

Comment:
@derekwaynecarr so it turns out to be pretty easy to reuse LimitRanger like this and then have it also validate after. Look good?